stdlib/public/core/Filter.swift

```diff
@@ -203,8 +205,28 @@ public struct LazyFilterCollection<
 
   // TODO: swift-3-indexing-model - add docs
   @warn_unused_result
-  public func next(index: Index) -> Index {
-    return LazyFilterIndex(base: _nextFiltered(index.base))
+  public func next(i: Index) -> Index {
+    return LazyFilterIndex(base: _nextFiltered(i.base))
+  }
+
+  // TODO: swift-3-indexing-model - add docs
+  @warn_unused_result
+  public func advance(i: Index, by n: IndexDistance) -> Index {
+    var index = i.base
+    for _ in 0..<n where index != _base.endIndex {
+      index = _nextFiltered(index)
+    }
+    return LazyFilterIndex(base: index)
+  }
+
+  // TODO: swift-3-indexing-model - add docs
+  @warn_unused_result
+  public func advance(i: Index, by n: IndexDistance, limit: Index) -> Index {
+    var index = i.base
+    for _ in 0..<n where index != _base.endIndex && index != limit.base {
```

Comment:
I think this will keep looping (but only for the sake of counting) even when we hit the limit.  I think the implementation modeled after `_advanceForward` would be appropriate here.